src/Data/Array.purs

```diff
@@ -1334,37 +1339,39 @@ slidingSizeStep size step array
 rangeWithStep :: Int -> Int -> Int -> Array Int
 rangeWithStep start end step = rangeWithStep' start end step identity
 
--- | Helper function to produce an array of elements like `rangeWithStep` with start, end and step, but immediatelly mapping over the result to work on one single mutable array.
+-- | Works just like rangeWithStep, but also uses a function to map each integer.
+-- | `rangeWithStep' start end step f` is the same as `map f $ rangeWithStep start end step`,
+-- | but without the extra intermediate array allocation.
 -- |
 -- | ```purescript
 -- | > rangeWithStep' 0 6 2 identity = [0,2,4,6]
 -- | > rangeWithStep' 0 6 2 (add 3) = [3,5,7,9]
+-- | > rangeWithStep' 0 (-6) (-2) (add 3) = [3,1,-1,-3]
 -- | ```
 rangeWithStep' :: forall t. Int -> Int -> Int -> (Int -> t) -> Array t
 rangeWithStep' start end step fn =
-  STA.run
-    ( do
-        let
-          isValid =
-            step /= 0
-              && if end >= start then
-                  step > 0
-                else
-                  step < 0
-
-          hasReachedEnd current =
-            if step > 0 then
-              current > end
-            else
-              current < end
-
-          helper current acc =
-            if hasReachedEnd current then
-              pure acc
-            else do
-              void $ STA.push (fn current) acc
-              helper (current + step) acc
-        arr <- STA.new
-        void $ helper start arr
-        pure arr
-    )
+  if not isValid 
+  then []
+  else STA.run
+   ( do
+      let
```

Comment:
Just a stylistic change. The parenthesis after the `ST.run` can be removed if you follow it with `do`
```purescript
else ST.run do
  let
    helper current acc = {- code... -}
    
  arr <- STA.new
```
Just to clarify, normally `ST.run $ codeBlock` will work. However, because of the Rank-2 type in `ST.run` (i.e. the `h` type parameter), using `$` in this way will always produce a compiler error. However, using `do` creates a new block, so the parenthesis are no longer needed.

Other than that, I think this is ready to merge

Reply:
I think I found that out the hard way when I tried to move the helper function out of the ST.run block. Never heard about rank-2 types before - at least not consciously ;-)